http-client/src/main/java/io/micronaut/http/client/DefaultHttpClient.java

```diff
@@ -1796,136 +1796,142 @@ private String getHostHeader(URI requestURI) {
 
             @Override
             protected void channelRead0(ChannelHandlerContext channelHandlerContext, FullHttpResponse fullResponse) {
-                HttpResponseStatus status = fullResponse.status();
-                int statusCode = status.code();
-                HttpStatus httpStatus;
-                try {
-                    httpStatus = HttpStatus.valueOf(statusCode);
-                } catch (IllegalArgumentException e) {
-                    emitter.onError(e);
-                    return;
-                }
-
                 try {
 
-                    HttpHeaders headers = fullResponse.headers();
-                    if (log.isTraceEnabled()) {
-                        log.trace("HTTP Client Response Received for Request: {} {}", request.getMethod(), request.getUri());
-                        log.trace("Status Code: {}", status);
-                        traceHeaders(headers);
-                        traceBody("Response", fullResponse.content());
+                    HttpResponseStatus status = fullResponse.status();
+                    int statusCode = status.code();
+                    HttpStatus httpStatus;
+                    try {
+                        httpStatus = HttpStatus.valueOf(statusCode);
+                    } catch (IllegalArgumentException e) {
+                        if (complete.compareAndSet(false, true)) {
+                            emitter.onError(e);
+                        } else if (LOG.isWarnEnabled()) {
+                            LOG.warn("Unsupported http status after handler completed: " + e.getMessage(), e);
+                        }
+                        return;
                     }
 
-                    // it is a redirect
-                    if (statusCode > 300 && statusCode < 400 && configuration.isFollowRedirects() && headers.contains(HttpHeaderNames.LOCATION)) {
-                        String location = headers.get(HttpHeaderNames.LOCATION);
-                        Flowable<io.micronaut.http.HttpResponse<O>> redirectedRequest = exchange(io.micronaut.http.HttpRequest.GET(location), bodyType);
-                        redirectedRequest.first(io.micronaut.http.HttpResponse.notFound())
-                                .subscribe((oHttpResponse, throwable) -> {
-                                    if (throwable != null) {
-                                        emitter.onError(throwable);
+                    try {
```

Comment:
Simply moving the code into the old try block should be sufficient yes?

Reply:
Not really - `HttpStatus httpStatus;` is used inside `catch` clause, so it cannot be declared inside. It could be, of course, declared and *initialized* outside (`HttpStatus httpStatus = null;`)... but the the `catch` has to handle this `null` value.

In the end, splitting the `catch` and `finally` is the simplest and safest solution.